mmcv/ops/csrc/pytorch/cuda/rans.cu

```diff
@@ -0,0 +1,259 @@
+    int begin_idx = GET_BEGIN_IDX(i, symbols.size(0), num_threads);
+    int end_idx = GET_END_IDX(i, symbols.size(0), num_threads);
+    cpu_threads.push_back(
+        std::thread(rans_encode_cpu_kernel, std::cref(rans_symbols_cpu),
+                    std::cref(results_cpu), begin_idx, end_idx,
+                    std::ref(outputs[i]), std::ref(nbytes[i])));
+  }
+
+  // start threads
+  for (auto &t : cpu_threads) {
+    t.join();
+  }
+
+  // set header
+  RansHeader header;
+  header.encode_with_cpu = true;
+  header.num_threads = num_threads;
+  header.nbytes_mode =
+      get_nbytes_mode(*std::max_element(nbytes.begin(), nbytes.end()));
+
+  // output encoded
+  std::string encoded = "";
+#pragma unroll
+  for (int i = 0; i < num_threads; i++) {
+    uint32_t *ptr =
+        outputs[i].data() + outputs[i].size() - nbytes[i] / sizeof(uint32_t);
+    encoded += std::string(reinterpret_cast<char *>(ptr), nbytes[i]);
+  }
+  // output nbytes
+  encoded += compress_nbytes(nbytes.data(), header.nbytes_mode, num_threads);
+  // output header
+  encoded += compress_rans_header(header);
+
+  return encoded;
+}
+
+Tensor RansDecodeWithIndexesCUDAKernelLauncher(const std::string &encoded,
+                                               const Tensor indexes,
+                                               const Tensor cdfs,
+                                               const Tensor cdfs_sizes,
+                                               const Tensor offsets) {
+  // param encoded: encoded symbols bitstream
+  // param indexes: (N, )
+  // param cdfs: (M, max_cdf_bin_size)
+  // param cdfs_sizes: (M, )
+  // param offsets: (M, )
+  check_rans_decode_input(encoded, indexes, cdfs, cdfs_sizes, offsets);
+
+  // allocate output
+  Tensor output = torch::zeros(indexes.sizes(), torch::kInt32);
+
+  // convert the input to vector
+  auto indexes_cpu = indexes.to(torch::kCPU);
+  auto cdfs_cpu = cdfs.to(torch::kCPU);
+  auto cdfs_sizes_cpu = cdfs_sizes.to(torch::kCPU);
+  auto offsets_cpu = offsets.to(torch::kCPU);
+  const auto indexes_vector = TENSOR_TO_VECTOR_1D(indexes_cpu, int32_t);
+  std::vector<std::vector<int32_t>> cdfs_vector(cdfs.size(0));
+  for (int i = 0; i < cdfs.size(0); i++) {
+    cdfs_vector[i] = TENSOR_TO_VECTOR_1D(cdfs_cpu[i], int32_t);
+  }
+  const auto cdfs_sizes_vector = TENSOR_TO_VECTOR_1D(cdfs_sizes_cpu, int32_t);
+  const auto offsets_vector = TENSOR_TO_VECTOR_1D(offsets_cpu, int32_t);
+
+  // decode header
+  RansHeader header = decompress_rans_header(
+      encoded.substr(encoded.length() - COMPRESS_HEADER_BYTES));
+  check_num_threads(header.num_threads);
+  int num_threads = header.num_threads;
+
+  // decode nbytes
+  uint32_t *nbytes = new uint32_t[num_threads];
+  int nbytes_bytes = num_threads * get_nbytes_size(header.nbytes_mode);
+  decompress_nbytes(
+      encoded.substr(encoded.length() - COMPRESS_HEADER_BYTES - nbytes_bytes),
+      nbytes, num_threads, header.nbytes_mode);
+
+  // multi-threading
```

Comment:
Please add guards and stream

Reply:
since the `stream` is not used in this function, could we only add the `guards`?
```python
  // set device
  at::cuda::CUDAGuard device_guard(symbols.device());
```

Reply:
The guard leads to a compile error, `symbols` has not been defined in this function.

Reply:
I have fixed this typo, and add stream into pmf_to_quantized_cdf_cuda_kernel.